src/Types/TxConstraints.purs

```diff
@@ -0,0 +1,370 @@
+  ( InputConstraint(..)
+  , OutputConstraint(..)
+  , TxConstraint(..)
+  , TxConstraints(..)
+  , addTxIn
+  , isSatisfiable
+  , modifiesUtxoSet
+  , mustBeSignedBy
+  , mustHashDatum
+  , mustIncludeDatum
+  , mustMintCurrency
+  , mustMintCurrencyWithRedeemer
+  , mustMintValue
+  , mustMintValueWithRedeemer
+  , mustPayToOtherScript
+  , mustPayToPubKey
+  , mustPayToTheScript
+  , mustProduceAtLeast
+  , mustProduceAtLeastTotal
+  , mustSatisfyAnyOf
+  , mustSpendAtLeast
+  , mustSpendAtLeastTotal
+  , mustSpendPubKeyOutput
+  , mustSpendScriptOutput
+  , mustValidateIn
+  , pubKeyPayments
+  , requiredDatums
+  , requiredMonetaryPolicies
+  , requiredSignatories
+  , singleton
+  ) where
+
+import Prelude hiding (join)
+
+import Data.Array as Array
+import Data.Array ((:), concat)
+import Data.Bifunctor (class Bifunctor)
+import Data.BigInt (BigInt)
+import Data.Foldable (class Foldable, any, foldl, foldMap, foldr, null)
+import Data.Generic.Rep (class Generic)
+import Data.Lattice (join)
+import Data.Map (fromFoldableWith, toUnfoldable)
+import Data.Maybe (Maybe(Just, Nothing))
+import Data.Newtype (class Newtype, over, unwrap)
+import Data.Show.Generic (genericShow)
+import Data.Tuple (uncurry)
+import Data.Tuple.Nested ((/\), type (/\))
+import Types.PlutusData
+  ( class ToData
+  , Datum(Datum)
+  , DatumHash
+  , toData
+  , unitRedeemer
+  )
+import Types.Interval (POSIXTimeRange, always, intersection, isEmpty)
+import Types.RedeemerTag (RedeemerTag(Mint))
+import Types.ScriptHash (MintingPolicyHash, ValidatorHash)
+import Types.Transaction (Redeemer)
+import Types.UnbalancedTransaction (PubKeyHash, TxOutRef)
+import Types.Value (TokenName, Value, currencyMPSHash, getNonAdaAsset, isZero)
+
+--------------------------------------------------------------------------------
+-- TxConstraints Type and related
+--------------------------------------------------------------------------------
+-- Taken from https://playground.plutus.iohkdev.io/doc/haddock/plutus-ledger-constraints/html/Ledger-Constraints.html
+
+-- | Constraints on transactions that want to spend script outputs
+data TxConstraint
+  = MustIncludeDatum Datum
+  | MustValidateIn POSIXTimeRange
+  | MustBeSignedBy PubKeyHash
+  | MustSpendAtLeast Value
+  | MustProduceAtLeast Value
+  | MustSpendPubKeyOutput TxOutRef
+  | MustSpendScriptOutput TxOutRef Redeemer
+  | MustMintValue MintingPolicyHash Redeemer TokenName BigInt
+  | MustPayToPubKey PubKeyHash Value
+  | MustPayToOtherScript ValidatorHash Datum Value
+  | MustHashDatum DatumHash Datum
+  | MustSatisfyAnyOf (Array (Array TxConstraint))
+
+derive instance eqTxConstraint :: Eq TxConstraint
+derive instance genericTxConstraint :: Generic TxConstraint _
+
+instance showTxConstraint :: Show TxConstraint where
+  show x = genericShow x
+
+newtype InputConstraint (i :: Type) = InputConstraint
+  { icRedeemer :: i
+  , icTxOutRef :: TxOutRef
+  }
+
+derive instance genericInputConstraint :: Generic (InputConstraint i) _
```

Comment:
do we need names for instances? they are noisy and usually bring no value :shrug:

Reply:
I can remove these tomorrow.